### DISCUSSION TITLE
Paginated API request for asset

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#17832] [Q&A] [answered]
Hi everyone, Dagster newbie here. What's the best way to handle a paginated API request for an asset?

Should I loop through all pages within the asset
```python
@asset
def messages():
  next_page = None
  has_more = True
  all_data = []

  while has_more:
    response = fetch_data_from_api(page = next_page)
    all_data = all_data.concat(response.data)
    next_page = response.next_page:
    if not next_page:
      has_more = False

  store_data(all_data)
```
Or is there a way to somehow store the next page cursor as context where Dagster knows to call `messages` again with new new cursor?

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1699471512217289?thread_ts=1699471512.217289&cid=C01U954MEER).*

**Comment by jamiedemaria** [comment, marked as answer]
Looping through all of the pages is certainly a good option/approach. if there is a reasonable way to partition the data (ie on a date field) then you could also look into partitioning your asset and only fetch + store data for a particular partition when that partition is materialized https://docs.dagster.io/concepts/partitions-schedules-sensors/partitions
